policy-controller/src/main.rs

```diff
@@ -111,17 +112,76 @@ async fn main() -> Result<()> {
 
-    // Run the admission controller
     if let Some(bind_addr) = admission_addr {
-        let (listen_addr, serve) = warp::serve(admission::routes(client))
-            .tls()
-            .cert_path("/var/run/linkerd/tls/tls.crt")
-            .key_path("/var/run/linkerd/tls/tls.key")
-            .bind_with_graceful_shutdown(bind_addr, async move {
-                let _ = drain_rx.signaled().await;
-            });
+        let tcp = TcpListener::bind(&bind_addr).await?;
+        let listen_addr = tcp.local_addr()?;
+        // Connection accept loop
+        let server = async move {
+            let http = hyper::server::conn::Http::new();
+            let handler = admission::Service { client };
+
+            loop {
+                let socket = match tcp.accept().await {
+                    Ok((socket, _)) => socket,
+                    Err(err) => {
+                        error!(%err, "Failed to accept connection");
+                        continue;
+                    }
+                };
+                let tls_cfg = {
+                    // Load public certificate.
+                    let certs = match load_certs("/var/run/linkerd/tls/tls.crt")
+                        .await
+                        .with_context(|| "failed to load certificate")
+                    {
+                        Ok(certs) => certs,
+                        Err(error) => {
+                            error!(%error);
+                            continue;
+                        }
+                    };
+                    // Load private key.
+                    let key = match load_private_key("/var/run/linkerd/tls/tls.key")
+                        .await
+                        .with_context(|| "failed to load private key")
+                    {
+                        Ok(key) => key,
+                        Err(error) => {
+                            error!(%error);
+                            continue;
+                        }
+                    };
+                    // Do not use client certificate authentication.
+                    let mut cfg = match rustls::ServerConfig::builder()
+                        .with_safe_defaults()
+                        .with_no_client_auth()
+                        .with_single_cert(certs, key)
+                    {
+                        Ok(cfg) => cfg,
+                        Err(error) => {
+                            error!(%error, "Failed to configure TLS");
+                            continue;
+                        }
+                    };
+                    // Configure ALPN to accept HTTP/2, HTTP/1.1 in that order.
+                    cfg.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
+                    sync::Arc::new(cfg)
+                };
+                let tls_acceptor = TlsAcceptor::from(tls_cfg);
+                let stream = match tls_acceptor.accept(socket).await {
```

Comment:
this `.await` (and the above where we read the files) will block the accept loop. so if this accept takes a while to complete or hangs indefinitely, we won't process new connections.